keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1274 @@
+the convenient property that its encoded form is always prefixed by
+`0xd9d9f7`. By design, this prefix is never found at the beginning of a JSON
+text and can be used as a “magic number” to distinguish the data format of a
+stored object at rest.
+
+To support decoding custom resources that have been stored as a mixture of JSON
+and CBOR, the CBOR serializer will implement RecognizingDecoder by checking for
+the prefix `0xd9d9f7`.
+
+Streaming responses (i.e. watches) will be serialized as CBOR Sequences. A [CBOR
+Sequence](https://www.rfc-editor.org/rfc/rfc8742.html) is a concatenation of
+zero or more CBOR data items, with no additional framing. This is effectively
+equivalent to the existing JSON stream serialization behavior and takes
+advantage of the property that, like JSON objects – and unlike Protobuf messages
+or non-object JSON documents, e.g. numbers – CBOR data items are
+self-delimiting.
+
+At the time of writing, watch events are encoded to a temporary buffer before
+being passed to the frame writer. Frame writers can also assume that the byte
+slice passed to each call of `Write` represents the complete contents of one
+frame. The Protobuf frame writer takes advantage of both in order to determine a
+frame's length prefix "for free". If this proposal were to require encoding
+events using the effectively length-prefixed approach described in [Optimizing
+CBOR Sequences for Skipping
+Elements](https://www.rfc-editor.org/rfc/rfc8742.html#name-optimizing-cbor-sequences-f),
+the CBOR frame writer would similarly need to know each event's encoded size.
+
+One useful property of a self-delimiting encoding is described [in the CBOR
+standard](https://www.rfc-editor.org/rfc/rfc8949.html#section-4.2.1-3.1):
+
+> the self-delimiting nature of the CBOR encoding means that there are no two
+> well-formed CBOR encoded data items where one is a prefix of the other
+
+In other words, CBOR (and the existing JSON framing) can stream directly to and
+from the wire without incurring additional copies on both sides of the
+connection. If an encoding fails or is otherwise not completely received on the
+other end, the fragment that _is_ received will not be well-formed and will
+produce a decode error.
+
+### Negotiation
+
+Proactive content negotiation will be supported for clients that want to receive
+CBOR-encoded responses using the MIME type “application/cbor” in the Accept
+request header. For compatibility with API servers that don’t support CBOR,
+clients should also accept “application/json” (with a lower quality factor) and
+choose the appropriate decoder based on the Content-Type response header.
+
+Similarly, streaming requests should use the MIME type for CBOR Sequences,
+“application/cbor-seq”.
+
+CBOR will not be a supported encoding for JSON Patches, JSON Merge Patches,
+Strategic Merge Patches, or Server-Side Apply configurations.
+
+Clients can send CBOR-encoded request bodies with the appropriate Content-Type
```

Comment:
I mentioned in a comment that got out-dated but you could possibly use the OpenAPI to check for accepted media-types.

Reply:
I've extended the proposal to allow client fallback to be bypassed using request media types from OpenAPI.